src/music/dmusic.cpp

```diff
@@ -108,69 +1115,55 @@ const char *MusicDriver_DMusic::Start(const char * const *parm)
 		caps.dwSize = sizeof(DMUS_PORTCAPS);
 
 		DEBUG(driver, 1, "Detected DirectMusic ports:");
-		for (int i = 0; music->EnumPort(i, &caps) == S_OK; i++) {
+		for (int i = 0; _music->EnumPort(i, &caps) == S_OK; i++) {
 			if (caps.dwClass == DMUS_PC_OUTPUTCLASS) {
 				/* Description is UNICODE even for ANSI build. */
-				DEBUG(driver, 1, " %d: %s%s", i, convert_from_fs(caps.wszDescription, desc, lengthof(desc)), i == port ? " (selected)" : "");
+				DEBUG(driver, 1, " %d: %s%s", i, convert_from_fs(caps.wszDescription, desc, lengthof(desc)), i == pIdx ? " (selected)" : "");
 			}
 		}
 	}
 
-	IDirectMusicPort *music_port = NULL; // NULL means 'use default port'.
-
-	if (port >= 0) {
+	GUID guidPort;
+	if (pIdx >= 0) {
 		/* Check if the passed port is a valid port. */
 		DMUS_PORTCAPS caps;
 		MemSetT(&caps, 0);
 		caps.dwSize = sizeof(DMUS_PORTCAPS);
-		if (FAILED(music->EnumPort(port, &caps))) return "Supplied port parameter is not a valid port";
+		if (FAILED(_music->EnumPort(pIdx, &caps))) return "Supplied port parameter is not a valid port";
 		if (caps.dwClass != DMUS_PC_OUTPUTCLASS) return "Supplied port parameter is not an output port";
+		guidPort = caps.guidPort;
+	} else {
+		if (FAILED(_music->GetDefaultPort(&guidPort))) return "Can't query default music port";
+	}
 
-		/* Create new port. */
-		DMUS_PORTPARAMS params;
-		MemSetT(&params, 0);
-		params.dwSize          = sizeof(DMUS_PORTPARAMS);
-		params.dwValidParams   = DMUS_PORTPARAMS_CHANNELGROUPS;
-		params.dwChannelGroups = 1;
+	/* Create new port. */
+	DMUS_PORTPARAMS params;
+	MemSetT(&params, 0);
+	params.dwSize = sizeof(DMUS_PORTPARAMS);
+	params.dwValidParams = DMUS_PORTPARAMS_CHANNELGROUPS;
+	params.dwChannelGroups = 1;
+	if (FAILED(_music->CreatePort(guidPort, &params, &_port, NULL))) return "Failed to create port";
+	/* Activate port. */
+	if (FAILED(_port->Activate(TRUE))) return "Failed to activate port";
 
-		if (FAILED(music->CreatePort(caps.guidPort, &params, &music_port, NULL))) {
-			return "Failed to create port";
-		}
+	/* Create playback buffer. */
+	DMUS_BUFFERDESC desc;
+	MemSetT(&desc, 0);
+	desc.dwSize = sizeof(DMUS_BUFFERDESC);
+	desc.guidBufferFormat = KSDATAFORMAT_SUBTYPE_DIRECTMUSIC;
+	desc.cbBuffer = 1024;
+	if (FAILED(_music->CreateMusicBuffer(&desc, &_buffer, NULL))) return "Failed to create music buffer";
 
-		/* Activate port. */
-		if (FAILED(music_port->Activate(TRUE))) {
-			music_port->Release();
-			return "Failed to activate port";
-		}
-	}
+	const char *dls = LoadDefaultDLSFile(GetDriverParam(parm, "dls"));
```

Comment:
This will likely fail on hardware synths, or really any time not using the default port. Should probably be guarded with `pIdx < 0`. (Haven't actually tested this yet.)

Reply:
`LoadDefaultDLSFile` is specifically checking for DLS support in the synth.